tokio/tests/fs_uring_read.rs

```diff
@@ -0,0 +1,194 @@
+fn current_rt() -> Box<dyn Fn() -> Runtime> {
+    Box::new(|| Builder::new_current_thread().enable_all().build().unwrap())
+}
+
+fn rt_combinations() -> Vec<Box<dyn Fn() -> Runtime>> {
+    vec![
+        current_rt(),
+        multi_rt(1),
+        multi_rt(2),
+        multi_rt(8),
+        multi_rt(64),
+        multi_rt(256),
+    ]
+}
+
+#[test]
+fn shutdown_runtime_while_performing_io_uring_ops() {
+    fn run(rt: Runtime) {
+        let (tx, rx) = mpsc::channel();
+        let (done_tx, done_rx) = mpsc::channel();
+
+        let (_tmp, path) = create_tmp_files(1);
+        rt.spawn(async move {
+            let path = path[0].clone();
+
+            // spawning a bunch of uring operations.
+            loop {
+                let path = path.clone();
+                tokio::spawn(async move {
+                    let bytes = read(path).await.unwrap();
+
+                    assert_eq!(bytes, vec![20; 1023]);
+                });
+
+                // Avoid busy looping.
+                tokio::task::yield_now().await;
+            }
+        });
+
+        std::thread::spawn(move || {
+            let rt: Runtime = rx.recv().unwrap();
+            rt.shutdown_timeout(Duration::from_millis(300));
+            done_tx.send(()).unwrap();
+        });
+
+        tx.send(rt).unwrap();
+        done_rx.recv().unwrap();
+    }
+
+    for rt in rt_combinations() {
+        run(rt());
+    }
+}
+
+#[test]
+fn read_many_files() {
+    fn run(rt: Runtime) {
+        const NUM_FILES: usize = 512;
+
+        let (_tmp_files, paths): (Vec<NamedTempFile>, Vec<PathBuf>) = create_tmp_files(NUM_FILES);
+
+        rt.block_on(async move {
+            let tracker = TaskTracker::new();
+
+            for i in 0..10_000 {
+                let path = paths.get(i % NUM_FILES).unwrap().clone();
+                tracker.spawn(async move {
+                    let bytes = read(path).await.unwrap();
+                    assert_eq!(bytes, vec![20; 1023]);
+                });
+            }
+            tracker.close();
+            tracker.wait().await;
+        });
+    }
+
+    for rt in rt_combinations() {
+        run(rt());
+    }
+}
+
+#[tokio::test]
+async fn read_small_large_files() {
+    let (_tmp, path) = create_large_temp_file();
+
+    let bytes = read(path).await.unwrap();
+
+    assert_eq!(bytes, create_buf(5000));
+
+    let (_tmp, path) = create_small_temp_file();
+
+    let bytes = read(path).await.unwrap();
+
+    assert_eq!(bytes, create_buf(20));
+}
+
+#[tokio::test]
+async fn cancel_op_future() {
+    let (_tmp_file, path): (Vec<NamedTempFile>, Vec<PathBuf>) = create_tmp_files(1);
+
+    let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
+    let handle = tokio::spawn(async move {
+        poll_fn(|cx| {
+            let fut = read(&path[0]);
+
+            // If io_uring is enabled (and not falling back to the thread pool),
+            // the first poll should return Pending.
+            let _pending = Box::pin(fut).poll_unpin(cx);
```

Comment:
This code currently drops the read future right away. If you want the future to stay around until the `handle.abort()` call, then you need to modify this code.
```suggestion
        let fut = read(&path[0]);
        tokio::pin!(fut);
        poll_fn(|cx| {
            // If io_uring is enabled (and not falling back to the thread pool),
            // the first poll should return Pending.
            let _pending = fut.as_mut().poll(cx);
```

Reply:
I have pushed a slightly different version of this